### PR TITLE
Create IndexDefintion objects from SQLAlchemy primary keys 

### PR DESCRIPTION
**Pull request by pmantica1**
In this diff, I create IndexDefinition objects from SQLAlchemy primary keys and add them to the SQLAlchemy SchemaGraph.

### CODE REVIEW

File: docs/source/advanced_features/macro_system.rst [hunk 1]
```diff
@@ -191,7 +191,7 @@ We can now observe the process of macro expansion in action:
 
 .. code:: python
 
-    from graphql_compiler.macros import perform_macro_expansion
+    from graphql_compiler.macros import get_schema_with_macros, perform_macro_expansion
```
Comment (obi1kenobi) [inline docs/source/advanced_features/macro_system.rst]: The changes to this file look great but they seem completely unrelated to this PR.

File: graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py [hunk 2]
```diff
@@ -65,3 +68,42 @@ def _get_edge_type_from_direct_edge(edge_name, direct_edge_descriptor):
...
+                        continue
+
+                if invalid_columns:
+                    continue
+
+                # Some SQL backends allow duplicate nulls in columns with unique indexes.
+                # However, other backends do not. Therefore, we set ignore_nulls=True to indicate
+                # that the backend may have duplicate nulls.
+                #
+                # Also, since SQLAlchemy also does not contain information about whether an index is
```
[the remark this answers is not in the document]
Comment (pmantica1) [reply graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py]: I think that we should add both unique constraints and unique indexes since a unique constraint is implemented as a unique index under the hood. I also think that it makes sense for a key constraint to be considered as an index.  

https://stackoverflow.com/questions/2434502/is-a-primary-key-automatically-an-index
https://www.mssqltips.com/sqlservertip/4270/difference-between-sql-server-unique-indexes-and-unique-constraints/

File: docs/source/advanced_features/macro_system.rst (hunk 1 above)
Comment (pmantica1) [reply docs/source/advanced_features/macro_system.rst]: Will remove.

File: graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py [hunk 3]
```diff
@@ -65,3 +68,42 @@ def _get_edge_type_from_direct_edge(edge_name, direct_edge_descriptor):
...
+    """Return the IndexDefinition objects corresponding to the indexes in the SQL tables."""
+    # TODO: Add indexes that do not correspond to neither the primary key nor the SQLAlchemy Table's
+    #       uniqueness constraints.
+    index_definitions = set()
+    for vertex_name, table in vertex_name_to_table.items():
+        for constraint in table.constraints:
+            if isinstance(constraint, (PrimaryKeyConstraint, UniqueConstraint)):
+                column_names = frozenset(column.name for column in constraint.columns)
+
+                # We ignore columns that have a SQL type with no matching GraphQL type.
```
[the remark this answers is not in the document]
Comment (pmantica1) [reply graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py]: I'll change this accordingly.

File: graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py [hunk 4]
```diff
@@ -65,3 +68,42 @@ def _get_edge_type_from_direct_edge(edge_name, direct_edge_descriptor):
...
+                invalid_columns = False
+                for name in column_names:
+                    if name not in vertex_types[vertex_name].properties:
+                        invalid_columns = True
+                        continue
```
[the remark this answers is not in the document]
Comment (pmantica1) [reply graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py]: Makes sense.

File: graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py [hunk 5]
```diff
@@ -65,3 +68,42 @@ def _get_edge_type_from_direct_edge(edge_name, direct_edge_descriptor):
...
+
+                if invalid_columns:
+                    continue
+
+                # Some SQL backends allow duplicate nulls in columns with unique indexes.
+                # However, other backends do not. Therefore, we set ignore_nulls=True to indicate
+                # that the backend may have duplicate nulls.
```
[the remark this answers is not in the document]
Comment (pmantica1) [reply graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py]: This would take me 2 hours to implement. Basically I would have to:
1. For each database, I would have to google this information.
2. Create a map of dialect to `ignored_nulls` value.
3. Figure out the behavior of what to do if the dialect is not in the list I included. (We also have to make sure this list is kept up-to-date if SQLAlchemy ever includes a new dialect). 

Since `ignore_nulls` does not have any use in the compiler I don't think this is worth it. Do you think it's worth it?

File: graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py [hunk 6]
```diff
@@ -65,3 +68,41 @@ def _get_edge_type_from_direct_edge(edge_name, direct_edge_descriptor):
...
+    # TODO(pmantica): Add non-primary key indexes.
+    index_definitions = set()
+    for vertex_name, table in vertex_name_to_table.items():
+        vertex_type = vertex_types[vertex_name]
+        for constraint in table.constraints:
+            # When a primary key constraint is created in a SQL database, an accompanying
+            # primary key index is also created. However, SQLAlchemy explicitly does not include
+            # primary key indexes in table.indexes. Therefore, we use the PrimaryKeyConstraint
+            # to create the IndexDefinition. The only issue with doing so is that technically
+            # the index name is incorrect.
```
Comment (bojanserafimov) [inline graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py]: Can you just not provide the name then? It's optional.

Comment (pmantica1) [reply graphql_compiler/schema_generation/sqlalchemy/schema_graph_builder.py]: Ok makes sense.
